### ISSUE TITLE
RFC (auth): separate initial and refresh auth

### ISSUE THREAD

**Issue by kadikraman** [urql-graphql/urql#2815]
## Summary

In the `auth-exchange`, we currently have one function - `getAuth` that handles fetching both initial auth as well as refreshing the auth in case of errors.

This proposal is to separate this function into two separate functions: one to handle the initial auth only, and the other to handle all subsequent calls.

These should be separated because currently users will have to keep track of their own auth state to know when to fetch the token from storage and when to trigger the refresh flow, and it's easy to make mistakes.

Furthermore, "get initial auth" and "refresh auth" are very different actions from a library user's point of view andit would reduce the mental load to keep them separated.

## Proposed Solution

Separate the `getAuth` into two functions:
1. `getInitialAuth`: called only once when the exchange first loads. It should be used to get any initial auth state (e.g. form a cookie or async storage)
2. `refreshAuth`: called when `willAuthError` returns true, or when a request fails with an auth error (as determined by `didAuthError`)

## Requirements

The proposed changes can be seen in this PR: https://github.com/urql-graphql/urql/pull/2814

**Comment by JoviDeCroock** [comment]
I am a fan of separating these out, I do have some thoughts about the naming in general as I was thinking of it in the sense of `setup` and `on(Auth)Error` as the refresh only gets called when `will-` or `didAuthError` returns `true`. I do recognise that `onError` might sound a bit severe 😅

**Comment by nandorojo** [comment]
I'd also like an option to `getAuth()` for every request. I previously did this by setting `willAuthError: () => true`, but that ends up sending double requests every time.

Firebase Auth doesn't expose the token expiration time, so there's no reliable way to know if auth will error ahead of time. However, it does expose a `getToken()` function which returns the existing token. I'd ideally like to call this prior to each request. However, I've had to revert back to `didAuthError()` and just intercept a custom error response.

**Comment by kadikraman** [comment]
Thanks for the feedback @nandorojo that's interesting. `willAuthError: () => true` would be the correct way to achieve calling `getAuth()` every time and it shouldn't be sending double requests. I'll have a look into it.

**Comment by nandorojo** [comment]
I see that this PR was meant to address just that: https://github.com/urql-graphql/urql/pull/1849

However, I've found that removing `willAuthError` is the only way around the double requests for me.

**Comment by kitten** [comment]
@nandorojo: I'm not quite sure this is subject to this issue, so if #1825 is an issue for you and you have a reproduction, I'll have to ask you to create a new issue please 😅

**Comment by kitten** [comment]
I gave this RFC some thought, and I think I also have a proposal for an alternative API. Essentially, I think the issue is that a "setup phase" and a "refresh phase" is hard to differentiate, but we're entering territory here of adding a _looooot_ of API 'hooks' for these different checks and phases, and I'd love to reduce them.

I think, we can potentially make the setup phase very explicit if we had a single entrypoint into the setup phase. I've also seen some confusion from people around the `authState`, since it's only local, but the timing aspect in it isn't actually as confusing to people than its mere naming.

So, I'd love to submit a second solution to this problem:

```js
// the first function is called on initial setup i.e. when the first request is made
// I'd pass utilities here, and potentially make a helper around `fetch` based on the client's options
authExchange(async ({ mutate }) => {
  // we're free here to initialise state as local state
  // or make extra requests already
  let token = localStorage.getItem('token');
  let refreshToken = localStorage.getItem('refreshToken');

  // we then return the config for future requests
  return {
    // this now takes no arguments but the operation
    willAuthError: (operation) => !!token && !!refreshState,

    // both of these only take a single argument, since type hints are sufficient
    // other arguments don't seem needed anymore
    didAuthError: (result) =>
      result.error.graphQLErrors.some(e => e.extensions?.code === 'FORBIDDEN'),

    // this now becomes a simple mapping function as well
    addAuthToOperation(operation) {
      if (!token) {
        return operation;
      }

      const fetchOptions =
        typeof operation.context.fetchOptions === 'function'
          ? operation.context.fetchOptions()
          : operation.context.fetchOptions || {};
    
      return makeOperation(operation.kind, operation, {
        ...operation.context,
        fetchOptions: {
          ...fetchOptions,
          headers: {
            ...fetchOptions.headers,
            Authorization: authState.token,
          },
        },
      });
    },

    // this now takes essentially no arguments at all and can return a Promise
    // because only some functions allow (in their type and impl.) to return a
    // promise, we automatically guide the user towards this function
    async refreshAuth() {
      const result = await mutate(refreshMutation, { refreshToken });

      if (result.data?.refreshLogin) {
        localStorage.setItem('token', token = result.data.refreshLogin.token);
        localStorage.setItem('refreshToken', refreshToken = result.data.refreshLogin.refreshToken);
      }

      localStorage.clear();
      logout();
    },
  };
}),
```

The idea here is essentially:

```js
authExchange(async (utilities) => {
  return {
    willAuthError: (operation) => false,
    didAuthError: (result) => false,
    addAuthToOperation: (operation) => operation,
    refreshAuth: async () => {/*...*/},
  };
}),
```

This allows users to:
- Rely on local state variables
- Imply that calling out to other async functions in `refreshAuth` or the setup call is fine
- Allow conditional returns and local functions to be more easily reusable
- Make it less awkward to share state with "external auth sources"

**Comment by kadikraman** [comment]
@kitten huge fan of getting rid of the `authState` - even in my own real world experience, it doesn't actually get used to store the JWT. It only gets used to determine whether `getAuth` needs to fetch the auth state from storage or not, but if we separate the `getAuth` to `getInitialAuth` and `refreshAuth`, this requirement goes away completely.

As for initialising the auth outside of the exchange. I'd argue that a lot of engineers would read this api as "`localStorage.getItem('token')` gets called every time the a request comes in" even though that's not the case. But if we added code examples with variable names that make this more explicit, that should be okay.

**Comment by kitten** [comment]
@kadikraman I also had to look a bit hard for this, but we planned a while ago to add a `mapExchange` in core that is called on every request and can return a promise or a non-promise mapped operation. That was meant as a replacement for `fetchOptions` and simple cases.

But I think we never got around to that. It would cover other cases though.

To show the overarching theme here, in general, I'd argue that we have some patterns where exchanges call a function repeatedly per operation, and other cases where we don't (Graphcache has a single config as well), so I'd say, we could even present all code samples as:

```js
authExchange(async function setup() {
})
```

and it'd be clear what the intention here is 🤔
In general, we'll also need to work on our TSDoc comments

**Comment by yankovalera** [comment]
I've spent a few months now heavily using the auth exchange and I'd love to see the `refreshAuth` / `getInitialAuth` as 2 functions, as it clearly explains the purpose of each one.

I think that having the setup() could work, but IMO it's less clear on the intention than having a `getInitialAuth` which is very explicit for developers using the exchange, and the change would be more simple to adopt for the old exchange users, as it's only splitting the old getAuth into 2 hooks `refreshAuth` / `getInitialAuth` (using a the new exchange would require minimal effort in existing codebase)

**Comment by kitten** [comment]
To be a little more precise, the intention of the setup function _isn't to make migration easier_. And that's a more or less important distinction for me, because I'm always very hesitant to optimise API design for few known cases. If we were to optimise this to either match up more with the old design, and put more weight on what feels more similar/easier for people who have seen (and gotten familiar with) the old API, then we risk accepting a subpar option 😅

The reason why I've basically found the `setup` elegant is that it gets around a problem that the `authExchange` already had in my opinion: API surface bloat.

We're talking about an exchange that could then potentially have the following callbacks:
- an initialiser (called once; async)
- a refresh (called repeatedly on error)
- will auth error (a pre-request, non-async predicate)
- did auth error (a post-request, non-async predicate)
- add auth to operation (a transform function)

And preserving all else, the concept of `authState` and having multiple object-parameters passed to all of these functions.

I've personally struggled in the past to explain people three concepts here:
- the auth state (and syncing this to external singleton state stores)
- the timing of getAuth/refresh
- why it's important to have will/did auth and have this option
- what happens to operation after or during any of these processes

So, I've taken the opportunity to propose the setup because it particularly helps to reduce this, by:
- removing the concept of auth state altogether
- mentally and obviously segregating the initialisation (since it has no name and is just a function returning the config, there's a lot of clarity in the timing and what happens; i.e. it's not a lifecycle step people have to think about)
- simplifying parameters in the process (because we'll have one place to pass utilities and all other predicates and the transform function only take and return one object from here on)

So, the proposal is a subtle distinction (after all, current `getAuth` functions can still be refactored in little time to a setup/refresh pattern or init/refresh pattern; I don't expect this to take more than 15min)
But, it does, at least in my view, reduce the amount of hidden concepts we've introduced here in the past

**Comment by vladshcherbin** [comment]
This examples would greatly simplify the flow!

When using cookies for token storage, you don't even need initial setup or `addAuthToOperation`, just `didAuthError` and `refreshAuth`. Current `auth` implementation is confusing to use in this case, but still works.

Love both proposals, @kitten more. Hopefully one of them will be added not so far in future ❤️

**Comment by kitten** [comment]
@kadikraman: I finally got around to implementing the initializer-based proposal. I know it was a little hard to visualize just from my notes, but I hope the PR will help, and if we can get this signed off on by all interested parties and get this ready, awesome! If not, happy to take more time or circle back on different solutions :v:

cc @JoviDeCroock
